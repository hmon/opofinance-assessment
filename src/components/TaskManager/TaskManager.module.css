.TaskManager {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  justify-content: flex-start;
  width: 100%;
  height: 100%;
  gap: 4rem;

  @media (max-width: 768px) {
    flex-direction: column;
    gap: 2rem;
  }

  .TasksWrapper {
    display: flex;
    flex-direction: column;
    align-items: stretch;
    width: 100%;
  }
}
