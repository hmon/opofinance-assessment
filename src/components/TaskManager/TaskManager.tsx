import React, { useMemo, useState } from "react";
import useTaskStore from "../../store/task.ts";
import TaskForm from "../TaskForm/TaskForm.tsx";
import { TextField } from "../TextField/TextField.tsx";
import styles from "./TaskManager.module.css";
import { TaskList } from "../TaskList/TaskList.tsx";
import useDebounce from "../../hooks/useDebounce.ts";

const TaskManager: React.FC = () => {
  const { tasks } = useTaskStore();
  const [filter, setFilter] = useState("All");
  const [searchQuery, setSearchQuery] = useState("");
  const debouncedSearchQuery = useDebounce(searchQuery, 500);

  const filteredTasks = useMemo(() => tasks.filter(
    (task) =>
      (filter === "All" || task.status === filter) &&
      task.title.toLowerCase().includes(debouncedSearchQuery)
  ), [debouncedSearchQuery, filter, tasks]);

  return (
    <div className={styles.TaskManager} data-testid="task-manager">
      <TaskForm/>

      <div className={styles.TasksWrapper} data-testid="tasks-wrapper">
        <div className="controls">
          <select onChange={(e) => setFilter(e.target.value)}>
            <option value="All">All</option>
            <option value="Pending">Pending</option>
            <option value="In Progress">In Progress</option>
            <option value="Completed">Completed</option>
          </select>

          <TextField
            type="text"
            placeholder="Search by title..."
            value={searchQuery}
            onChange={(e) => setSearchQuery(e.target.value)}
          />
        </div>

        <TaskList tasks={filteredTasks}/>
      </div>
    </div>
  );
};

export default TaskManager;
