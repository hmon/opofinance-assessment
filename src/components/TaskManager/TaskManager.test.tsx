import { render, screen } from '@testing-library/react';
import TaskManager from './TaskManager';

describe('TaskManager Component', () => {
  it('renders TaskManager component', () => {
    render(<TaskManager />);
    const taskManagerElement = screen.getByTestId('task-manager');
    expect(taskManagerElement).toBeInTheDocument();
    expect(taskManagerElement).toMatchSnapshot();
  });

  test('renders TasksWrapper inside TaskManager', () => {
    render(<TaskManager />);
    const tasksWrapperElement = screen.getByTestId('tasks-wrapper');
    expect(tasksWrapperElement).toBeInTheDocument();
  });

  test('renders TaskList inside TasksWrapper', () => {
    render(<TaskManager />);
    const taskListElement = screen.getByTestId('task-list');
    expect(taskListElement).toBeInTheDocument();
  });
});
