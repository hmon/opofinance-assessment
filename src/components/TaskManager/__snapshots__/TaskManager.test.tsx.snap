// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskManager Component > renders TaskManager component 1`] = `
<div
  class="_TaskManager_0c2a9a"
  data-testid="task-manager"
>
  <form
    class="_TaskForm_e8b683"
    data-testid="task-form"
  >
    <div
      class="_TextField_da86b6"
    >
      <label
        for=":r0:"
      >
        Title
      </label>
      <input
        class="_input_da86b6"
        id=":r0:"
        value=""
      />
    </div>
    <div
      class="_TextField_da86b6"
    >
      <label
        for=":r1:"
      >
        Description
      </label>
      <input
        class="_input_da86b6"
        id=":r1:"
        value=""
      />
    </div>
    <div
      class="_TextField_da86b6"
    >
      <label
        for=":r2:"
      >
        Due Date
      </label>
      <input
        class="_input_da86b6"
        id=":r2:"
        type="date"
        value=""
      />
    </div>
    <button
      class="_Button_2a1efe"
      type="submit"
    >
      Add Task
    </button>
  </form>
  <div
    class="_TasksWrapper_0c2a9a"
    data-testid="tasks-wrapper"
  >
    <div
      class="controls"
    >
      <select>
        <option
          value="All"
        >
          All
        </option>
        <option
          value="Pending"
        >
          Pending
        </option>
        <option
          value="In Progress"
        >
          In Progress
        </option>
        <option
          value="Completed"
        >
          Completed
        </option>
      </select>
      <div
        class="_TextField_da86b6"
      >
        <input
          class="_input_da86b6"
          id=":r3:"
          placeholder="Search by title..."
          type="text"
          value=""
        />
      </div>
    </div>
    <ul
      class="_TaskList_e8740a"
      data-testid="task-list"
    >
      <div
        data-virtuoso-scroller="true"
        style="position: relative; height: 100%;"
      >
        <div
          data-viewport-type="window"
          style="width: 100%; height: 100%; position: absolute; top: 0px;"
        >
          <div
            data-testid="virtuoso-item-list"
            style="box-sizing: border-box; margin-top: 0px; padding-top: 0px; padding-bottom: 0px;"
          />
        </div>
      </div>
    </ul>
  </div>
</div>
`;
